src/Simplex/Messaging/Agent.hs

```diff
@@ -1979,6 +2134,48 @@ processSMPTransmission c@AgentClient {smpClients, subQ} (tSess@(_, srv, _), v, s
           | internalPrevMsgHash /= receivedPrevMsgHash = MsgError MsgBadHash
           | otherwise = MsgError MsgDuplicate -- this case is not possible
 
+withSwitchedRQ :: AgentMonad m => AgentClient -> RcvQueue -> RcvSwitchStatus -> (DB.Connection -> RcvQueue -> IO a) -> m (Either QueueCheckError a)
+withSwitchedRQ c rq expected =
+  withExistingRQ c rq (\RcvQueue {switchStatus} -> switchStatus == Just expected)
+
+withExistingRQ :: AgentMonad m => AgentClient -> RcvQueue -> (RcvQueue -> Bool) -> (DB.Connection -> RcvQueue -> IO a) -> m (Either QueueCheckError a)
+withExistingRQ c RcvQueue {connId, dbQueueId} qp action =
```

Comment:
why are we reading it here again if we already have it? should it just be used?

Reply:
this function seems completely redundant - it reads what it already have and then performs the passed action. All that's needed is to call the action from the places where it's called, and this function can be then removed.

Reply:
it re-reads the queue to check there were no concurrent database changes. Again, may be unnecessary due to conn lock but I kept it as a separate fail-safe